### ISSUE TITLE
[Bug] RMSprop 的 grad_clip 實作為 per-parameter clipping，與 Adam/AdamW 有同樣的問題

### ISSUE THREAD

**Issue by s9213712** [s9213712/minicnn#46]
## 問題描述

`src/minicnn/optim/rmsprop.py` 中的梯度裁剪：

```python
def step(self):
    for i, p in enumerate(self.params):
        if p.grad is None:
            continue
        grad = p.grad + self.weight_decay * p.data if self.weight_decay else p.grad
        if self.grad_clip > 0.0:
            norm = float(np.linalg.norm(grad))  # ← 單一參數的 L2 norm
            if norm > self.grad_clip:
                grad = grad * (self.grad_clip / norm)  # ← 各參數獨立裁剪
```

與 `optim/adam.py` 和 `optim/adamw.py` 的實作方式相同，都是 **per-parameter L2 norm clipping**，而非業界標準的 **global gradient norm clipping**（clip the total L2 norm across all parameters）。

## 問題

四個優化器（SGD、Adam、AdamW、RMSprop）全部使用同一個有問題的 per-parameter clipping 實作。這是框架層面的設計不一致：

- PyTorch 提供 `torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm)` → global norm clipping
- 各優化器的 `grad_clip` 參數 → 應為 max_norm，但實際上是 per-parameter 的

此問題已在 issue #32 中針對 Adam 描述，本 issue 追蹤 RMSprop 的相同問題，以確保修復時所有優化器同步更新。

## 建議修復

在 `optim/optimizer.py` 基類中實作 global norm clipping，由子類呼叫：

```python
class Optimizer:
    def _apply_grad_clip(self) -> None:
        if self.grad_clip <= 0.0:
            return
        grads = [p.grad for p in self.params if p.grad is not None]
        if not grads:
            return
        total_norm = float(np.sqrt(sum(np.linalg.norm(g) ** 2 for g in grads)))
        if total_norm > self.grad_clip:
            scale = self.grad_clip / (total_norm + 1e-6)
            for p in self.params:
                if p.grad is not None:
                    p.grad = p.grad * scale
```
